File: messageformat/schema_test.go
package messageformat

import (
	"bytes"
	"encoding/json"
	"os"
	"sync"
	"testing"

	"github.com/santhosh-tekuri/jsonschema/v6"
)

const dataModelSchemaPath = "testdata/unicode/data-model/message.schema.json"

var (
	schemaOnce sync.Once
	schema     *jsonschema.Schema
	schemaErr  error
)

// dataModelSchema compiles the vendored MF2 data model JSON Schema, the wire
// contract every serialized message must satisfy.
func dataModelSchema(t *testing.T) *jsonschema.Schema {
	t.Helper()
	schemaOnce.Do(func() {
		f, err := os.Open(dataModelSchemaPath)
		if err != nil {
			schemaErr = err
			return
		}
		defer func() { _ = f.Close() }()
		doc, err := jsonschema.UnmarshalJSON(f)
		if err != nil {
			schemaErr = err
			return
		}
		c := jsonschema.NewCompiler()
		if err := c.AddResource("message.schema.json", doc); err != nil {
			schemaErr = err
			return
		}
		schema, schemaErr = c.Compile("message.schema.json")
	})
	if schemaErr != nil {
		t.Fatalf("compile data model schema: %v", schemaErr)
	}
	return schema
}

// assertMatchesSchema validates a JSON document against the data model schema.
func assertMatchesSchema(t *testing.T, doc []byte) {
	t.Helper()
	inst, err := jsonschema.UnmarshalJSON(bytes.NewReader(doc))
	if err != nil {
		t.Fatalf("decode instance: %v", err)
	}
	if err := dataModelSchema(t).Validate(inst); err != nil {
		t.Errorf("document does not match the MF2 data model schema: %v\n%s", err, doc)
	}
}

// assertMessageMatchesSchema encodes msg and validates it against the schema.
func assertMessageMatchesSchema(t *testing.T, msg Message) {
	t.Helper()
	doc, err := json.Marshal(msg)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	assertMatchesSchema(t, doc)
}

func TestSchemaRejectsInvalidDocument(t *testing.T) {
	inst, err := jsonschema.UnmarshalJSON(bytes.NewReader([]byte(`{"type":"message","pattern":[]}`)))
	if err != nil {
		t.Fatal(err)
	}
	if dataModelSchema(t).Validate(inst) == nil {
		t.Fatal("schema accepted a message without declarations; the validator is not wired up")
	}
}

func TestEncodedMessagesMatchSchema(t *testing.T) {
	msgs := []Message{
		{Type: PatternMessageType},
		{Type: SelectMessageType},
		{
			Type: SelectMessageType,
			Declarations: []Declaration{{
				Type:  InputDeclaration,
				Name:  "n",
				Value: Expression{Arg: VariableRef{Name: "n"}, Function: &FunctionRef{Name: "number"}},
			}},
			Selectors: []VariableRef{{Name: "n"}},
			Variants: []Variant{
				{Keys: []VariantKey{{Value: "one"}}, Value: Pattern{Expression{Arg: VariableRef{Name: "n"}}, Text(" Datei")}},
				{Keys: []VariantKey{{Catchall: true}}, Value: Pattern{Markup{Kind: MarkupStandalone, Name: "br", Attributes: Attributes{"x": nil}}}},
			},
		},
	}
	for _, m := range msgs {
		assertMessageMatchesSchema(t, m)
	}
}

File: messageformat/js/src/schema.test.ts
import { readFileSync } from "node:fs";
import { Ajv } from "ajv";
import { describe, expect, it } from "vitest";
import { isMessage, messageSchema } from "./schema.js";
import { parseMF2 } from "./mf2.js";
import { caseKind, messageSchemaPath, suiteCases } from "./testing/index.js";

const ajv = new Ajv({ strict: false });
const jsonSchema = JSON.parse(readFileSync(messageSchemaPath, "utf8")) as object;
const validateJson = ajv.compile(jsonSchema);

const parsable = suiteCases().filter(
  (tc) => caseKind(tc) !== "syntax-error" && caseKind(tc) !== "data-model-error",
);

describe("messageSchema agrees with message.schema.json", () => {
  it.each(parsable.map((tc) => [tc.src, tc] as const))("accepts %j", (_src, tc) => {
    const msg = parseMF2(tc.src);
    expect(validateJson(msg), JSON.stringify(validateJson.errors)).toBe(true);
    expect(messageSchema.safeParse(msg).success).toBe(true);
    expect(isMessage(msg)).toBe(true);
  });

  const lit = { type: "literal", value: "x" };
  const variable = { type: "variable", name: "x" };
  const expr = { type: "expression", arg: variable };
  const invalid: Array<[string, unknown]> = [
    ["null", null],
    ["unknown message type", { type: "nope", declarations: [], pattern: [] }],
    ["missing declarations", { type: "message", pattern: [] }],
    ["number in pattern", { type: "message", declarations: [], pattern: [1] }],
    ["expression without arg or function", {
      type: "message", declarations: [], pattern: [{ type: "expression" }],
    }],
    ["options as a Map", {
      type: "message", declarations: [], pattern: [{
        type: "expression", arg: lit,
        function: { type: "function", name: "number", options: new Map() },
      }],
    }],
    ["non-string literal", {
      type: "message", declarations: [], pattern: [{ type: "expression", arg: { type: "literal", value: 1 } }],
    }],
    ["input declaration with literal arg", {
      type: "message", pattern: [],
      declarations: [{ type: "input", name: "x", value: { type: "expression", arg: lit } }],
    }],
    ["attribute false", {
      type: "message", declarations: [], pattern: [{ ...expr, attributes: { a: false } }],
    }],
    ["markup kind", {
      type: "message", declarations: [], pattern: [{ type: "markup", kind: "self", name: "b" }],
    }],
    ["literal selector", {
      type: "select", declarations: [], selectors: [lit], variants: [],
    }],
    ["variant key of wrong type", {
      type: "select", declarations: [], selectors: [variable],
      variants: [{ keys: [variable], value: [] }],
    }],
    ["variant without value", {
      type: "select", declarations: [], selectors: [variable], variants: [{ keys: [] }],
    }],
  ];
  it.each(invalid)("rejects %s", (_name, value) => {
    const json = value instanceof Object ? JSON.parse(JSON.stringify(value)) as unknown : value;
    // A Map serializes to {} which the JSON Schema accepts; zod must reject the Map itself.
    if (!(_name as string).includes("Map")) expect(validateJson(json)).toBe(false);
    expect(messageSchema.safeParse(value).success).toBe(false);
    expect(isMessage(value)).toBe(false);
  });
});
